server/bin/pbench-audit-server.py

```diff
@@ -0,0 +1,590 @@
+                try:
+                    item_p = item_path.resolve(strict=True)
+                except FileNotFoundError:
+                    print(
+                        f"{item_path.name}_path, '{item_path}', does not resolve to a real location"
+                    )
+                    return 1
+                symlink_item = f"{item_path.name} -> {item_p}"
+                hier.add_unexpected_controllers(
+                    Hierarchy.UNEXPECTED_SYMLINKS, controller, symlink_item
+                )
+            elif all(
+                [
+                    item_path.is_file(),
+                    not item_path.name.endswith(".tar.xz"),
+                    not item_path.name.endswith(".tar.xz.md5"),
+                ]
+            ):
+                hier.add_unexpected_controllers(
+                    Hierarchy.UNEXPECTED_OBJECTS, controller, item_path.name
+                )
+            elif item_path.is_file() and (
+                item_path.name.endswith(".tar.xz")
+                or item_path.name.endswith(".tar.xz.md5")
+            ):
+                hier.add_tarballs(controller)
+            else:
+                print(
+                    f"{item_path} item should have been handled by the above mentioned conditions. "
+                    f"It is an unexpected item which should not have occured, "
+                    f"leading to an inappropriate condition like a hidden directory."
+                )
+        verify_subdirs(hier, controller, controller_subdir)
+        verify_prefixes(hier, controller)
+
+    return 0
+
+
+# Incoming Hierarchy
+def verify_tar_dirs(ihier, tarball_dirs, tblist, controller):
+    for tb in tarball_dirs:
+        if tb.endswith("unpack"):
+            tar = tb[:-7]
+            tar = f"{tar}.tar.xz"
+            val = Hierarchy.INVALID_UNPACKING_DIRS
+        else:
+            tar = f"{tb}.tar.xz"
+            val = Hierarchy.INVALID_TB_DIRS
+        tarfile = Path(ihier.config.ARCHIVE, controller, tar)
+        if tarfile.exists():
+            with tarfile.open(mode="r") as f:
+                try:
+                    file_content = f.read(1)
+                    if file_content:
+                        continue
+                except Exception:
+                    pass
+        else:
+            tblist(val, controller, tb)
+    return 0
+
+
+# 'ihier' is IncomingHierarchy class object
+def verify_incoming(ihier, verifylist):
+
+    for controller in verifylist:
+        ihier.add_controller(controller)
+        direct_entries = Path(ihier.config.INCOMING, controller).glob("*")
+        tarball_dirs = list()
+        unpacking_tarball_dirs = list()
+
+        if not Path(ihier.config.ARCHIVE, controller).is_dir():
+            # Skip incoming controller directories that don't have an $ARCHIVE
+            # directory, handled in another part of the audit.
+            continue
+
+        for dir_path in direct_entries:
+            if (
```

Comment:
This `if` needs refactoring in order to avoid doing the `listdir` unnecessarily (and even worse, doing it multiple times). Something like this:
```
   if dir_path.is_dir():
       if endswith(".unpack"):
           # we caught an unpack in progress - skip it
       else:
           if len(os.listdir(dir_path)) == 0:
               add it to empty tarball dirs
           else:
               add it to tarball_dirs
   elif dir_path.is_symlink():
       ihier.add_unexpected_controllers(Hierarchy.TARBALL_LINKS, controller, dir_path.name)
   else:
       # impossible: neither dir nor symlink - capture an error for reporting later.

Reply:
We seems to be catching controllers in impossible else condition

Reply:
We seem to be catching controllers in impossible else condition